exporters/prometheus/src/prometheus_exporter.cc

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "opentelemetry/exporters/prometheus/prometheus_exporter.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace exporter
+{
+namespace prometheus
+{
+/**
+ * Constructor - binds an exposer and collector to the exporter
+ * @param address: an address for an exposer that exposes
+ *  an HTTP endpoint for the exporter to connect to
+ */
+PrometheusExporter::PrometheusExporter(std::string &address) : is_shutdown_(false)
+{
+  exposer_   = std::unique_ptr<::prometheus::Exposer>(new ::prometheus::Exposer{address});
+  collector_ = std::shared_ptr<PrometheusCollector>(new PrometheusCollector);
+
+  exposer_->RegisterCollectable(collector_);
+}
+
+/**
+ * PrometheusExporter constructor with no parameters
+ * Used for testing only
+ */
+PrometheusExporter::PrometheusExporter() : is_shutdown_(false)
+{
+  exposer_   = nullptr;
+  collector_ = std::unique_ptr<PrometheusCollector>(new PrometheusCollector);
+}
+
+/**
+ * Exports a batch of Metric Records.
+ * @param records: a collection of records to export
+ * @return: returns a ReturnCode detailing a success, or type of failure
+ */
+sdk::metrics::ExportResult PrometheusExporter::Export(
+    const std::vector<sdk::metrics::Record> &records) noexcept
+{
+  if (is_shutdown_)
+  {
+    return sdk::metrics::ExportResult::kFailure;
+  }
+  else if (records.empty())
+  {
+    return sdk::metrics::ExportResult::kFailureInvalidArgument;
+  }
+  else if (collector_->GetCollection().size() + records.size() > collector_->GetMaxCollectionSize())
+  {
+    return sdk::metrics::ExportResult::kFailureFull;
+  }
+  else
+  {
+    collector_->AddMetricData(records);
+    return sdk::metrics::ExportResult::kSuccess;
+  }
+}
+
+/**
+ * Shuts down the exporter and does cleanup.
+ * Since Prometheus is a pull based interface,
+ * we cannot serve data remaining in the intermediate
+ * collection to to client an HTTP request being sent,
+ * so we flush the data.
+ */
+void PrometheusExporter::Shutdown() noexcept
+{
+  is_shutdown_ = true;
```

Comment:
Is Shutdown called from a different thread? Should there be a lock here?

Reply:
After looking into it, it seems as though Shutdown() will never be called by the controller in the SDK, since the Metrics Exporter Interface does not require a shutdown function - is there any benefit to keeping this in? If yes, then it seems we might need a lock here.

Reply:
What does the specification say about exporter shutdowns?

Reply:
It seems that it is not required for metrics exporters currently, but is required for span exporters. [Here](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#shutdown-1) is the requirement for span exporters, but looking at the shutdown function in the [ostream metrics exporter](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/docs/cpp-ostream-exporter-design.md#shutdown-1), it is mentioned but not required.

Reply:
At some point this might be nice to have. I think Ryan was saying that e.g. Envoy wants to be able to handle SIGHUP by completely shutting everything down, changing the config, and then bringing it back up.

Maybe the right answer is to punt on Shutdown for now and just leave a TODO? Idk.

Reply:
Maybe we can comment the shutdown function out, and add a TODO comment once the metrics specification adds a requirement for a shutdown?